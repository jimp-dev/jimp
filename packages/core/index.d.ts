export function addConstants(constants: [string, string | number], jimpInstance?: Jimp): void;
export function addJimpMethods(methods: [string, Function], jimpInstance?: Jimp): void;
export function jimpEvMethod(methodName: string, evName: string, method: Function): void;
export function jimpEvChange(methodName: string, method: Function): void;
export function addType(mime: string, extensions: string[]): void;

interface BaseJimp {
  prototype: Jimp;
  // Constants
  AUTO: -1;
  // blend modes
  BLEND_SOURCE_OVER: string;
  BLEND_DESTINATION_OVER: string;
  BLEND_MULTIPLY: string;
  BLEND_SCREEN: string;
  BLEND_OVERLAY: string;
  BLEND_DARKEN: string;
  BLEND_LIGHTEN: string;
  BLEND_HARDLIGHT: string;
  BLEND_DIFFERENCE: string;
  BLEND_EXCLUSION: string;
  // Align modes for cover, contain, bit masks
  HORIZONTAL_ALIGN_LEFT: 1;
  HORIZONTAL_ALIGN_CENTER: 2;
  HORIZONTAL_ALIGN_RIGHT: 4;
  VERTICAL_ALIGN_TOP: 8;
  VERTICAL_ALIGN_MIDDLE: 16;
  VERTICAL_ALIGN_BOTTOM: 32;
  // Edge Handling
  EDGE_EXTEND: 1;
  EDGE_WRAP: 2;
  EDGE_CROP: 3;
  // Properties
  bitmap: Bitmap;
  _quality: number;
  _deflateLevel: number;
  _deflateStrategy: number;
  _filterType: number;
  _rgba: boolean;
  _background: number;
  _originalMime: string;
  // Constructors
  new(path: string, cb?: ImageCallback): Jimp;
  new(urlOptions: URLOptions, cb?: ImageCallback): Jimp;
  new(image: Jimp, cb?: ImageCallback): Jimp;
  new(data: Buffer, cb?: ImageCallback): Jimp;
  new(data: Bitmap, cb?: ImageCallback): Jimp;
  new(w: number, h: number, cb?: ImageCallback): Jimp;
  new(
    w: number,
    h: number,
    background?: number | string,
    cb?: ImageCallback
  ): Jimp;
  // For custom constructors when using Jimp.appendConstructorOption
  new(...args: any[]): Jimp;
  // Methods
  on<T extends ListenableName>(
    event: T,
    cb: (data: ListenerData<T>) => any
  ): any;
  parseBitmap(
    data: Buffer,
    path: string | null | undefined,
    cb?: ImageCallback
  ): void;
  hasAlpha(): boolean;
  getHeight(): number;
  getWidth(): number;
  inspect(): string;
  toString(): string;
  getMIME(): string;
  getExtension(): string;
  distanceFromHash(hash: string): number;
  write(path: string, cb?: ImageCallback): this;
  writeAsync(path: string): Promise<Jimp>;
  deflateLevel(l: number, cb?: ImageCallback): this;
  deflateStrategy(s: number, cb?: ImageCallback): this;
  colorType(s: number, cb?: ImageCallback): this;
  filterType(f: number, cb?: ImageCallback): this;
  rgba(bool: boolean, cb?: ImageCallback): this;
  quality(n: number, cb?: ImageCallback): this;
  getBase64(mime: string, cb: GenericCallback<string, any, this>): this;
  getBase64Async(mime: string): Promise<string>;
  hash(cb?: GenericCallback<string, any, this>): string;
  hash(
    base: number | null | undefined,
    cb?: GenericCallback<string, any, this>
  ): string;
  getBuffer(mime: string, cb: GenericCallback<Buffer>): this;
  getBufferAsync(mime: string): Promise<Buffer>;
  getPixelIndex(
    x: number,
    y: number,
    cb?: GenericCallback<number, any, this>
  ): number;
  getPixelIndex(
    x: number,
    y: number,
    edgeHandling: string,
    cb?: GenericCallback<number, any, this>
  ): number;
  getPixelColor(
    x: number,
    y: number,
    cb?: GenericCallback<number, any, this>
  ): number;
  getPixelColour(
    x: number,
    y: number,
    cb?: GenericCallback<number, any, this>
  ): number;
  setPixelColor(hex: number, x: number, y: number, cb?: ImageCallback): this;
  setPixelColour(hex: number, x: number, y: number, cb?: ImageCallback): this;
  clone(cb?: ImageCallback): this;
  cloneQuiet(cb?: ImageCallback): this;
  background(hex: number, cb?: ImageCallback): this;
  backgroundQuiet(hex: number, cb?: ImageCallback): this;
  scan(
    x: number,
    y: number,
    w: number,
    h: number,
    f: (this: this, x: number, y: number, idx: number) => any,
    cb?: ImageCallback
  ): this;
  scanQuiet(
    x: number,
    y: number,
    w: number,
    h: number,
    f: (this: this, x: number, y: number, idx: number) => any,
    cb?: ImageCallback
  ): this;
  scanIterator(
    x: number,
    y: number,
    w: number,
    h: number
  ): IterableIterator<{x: number, y: number, idx: number, image: Jimp}>;

  // Effect methods
  composite(
    src: Jimp,
    x: number,
    y: number,
    options?: BlendMode,
    cb?: ImageCallback
  ): this;

  // Functions
  appendConstructorOption<T extends any[]>(
    name: string,
    test: (...args: T[]) => boolean,
    run: (
      this: Jimp,
      resolve: (jimp: Jimp) => any,
      reject: (reason: Error) => any,
      ...args: T[]
    ) => any
  ): void;
  read(path: string): Promise<Jimp>;
  read(image: Jimp): Promise<Jimp>;
  read(data: Buffer): Promise<Jimp>;
  read(w: number, h: number, background?: number | string): Promise<Jimp>;
  create(path: string): Promise<Jimp>;
  create(image: Jimp): Promise<Jimp>;
  create(data: Buffer): Promise<Jimp>;
  create(w: number, h: number, background?: number | string): Promise<Jimp>;
  rgbaToInt(
    r: number,
    g: number,
    b: number,
    a: number,
    cb: GenericCallback<number, any, Jimp>
  ): number;
  intToRGBA(i: number, cb?: GenericCallback<RGBA>): RGBA;
  cssColorToHex(cssColor: string): number;
  limit255(n: number): number;
  diff(
    img1: Jimp,
    img2: Jimp,
    threshold?: number
  ): {
    percent: number;
    image: Jimp;
  };
  distance(img1: Jimp, img2: Jimp): number;
  compareHashes(hash1: string, hash2: string): number;
  colorDiff(rgba1: RGB, rgba2: RGB): number;
  colorDiff(rgba1: RGBA, rgba2: RGBA): number;
}

export interface Image {
  bitmap: Bitmap;
}

// This must be exported to fix the "index signature missing" error
export interface IllformedPlugin {
  class?: never;
  constants?: never;
  [classFunc: string]: Function
}

export type DecoderFn = (data: Buffer) => Bitmap
export type EncoderFn<ImageType extends Image = Image> = (image: ImageType) => Buffer

interface WellFormedPlugin<ImageType extends Image = Image> {
  mime?: {
    [MIME_TYPE: string]: string[];
  };
  hasAlpha?: {
    [MIME_SPECIAL: string]: boolean;
  };
  constants?: {
    // Contants to assign to the Jimp instance
    [MIME_SPECIAL: string]: any;
  };
  decoders?: {
    [MIME_TYPE: string]: DecoderFn;
  };
  encoders?: {
    // Jimp Image
    [MIME_TYPE: string]: EncoderFn<ImageType>;
  };
  // Extend the Jimp class with the following constants, etc
  class?: any;
}

type ClassOrConstantPlugin<T extends Image> = WellFormedPlugin<T> & (
  Required<Pick<WellFormedPlugin<T>, 'class'>> | Required<Pick<WellFormedPlugin<T>, 'constants'>>
  );

// A Jimp type requires mime, but not class
export type JimpType<T extends Image = Image> = WellFormedPlugin<T> & Required<Pick<WellFormedPlugin<T>, 'mime'>>;

// Jimp plugin either MUST have class OR constant or be illformed
export type JimpPlugin<T extends Image = Image> = ClassOrConstantPlugin<T> | IllformedPlugin;

export type PluginFunction = () => JimpPlugin;
export type TypeFunction = () => JimpType;

// This is required as providing type arrays gives a union of all the generic
// types in the array rather than an intersection
type UnionToIntersection<U> =
  (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never

// The values to be extracted from a WellFormedPlugin to put onto the Jimp instance
type WellFormedValues<T extends WellFormedPlugin> = T['class'] & T['constants'];

// Jimp generic to be able to put plugins and types into, thus allowing
// `configure` from `@jimp/custom` to have proper typings
export type Jimp<T extends JimpType | undefined = undefined, P extends JimpPlugin | undefined = undefined> =
  BaseJimp
  & UnionToIntersection<WellFormedValues<T>>
  & UnionToIntersection<(P extends WellFormedPlugin ? WellFormedValues<P> : P)>

export type GenericCallback<T, U = any, TThis = any> = (
  this: TThis,
  err: Error | null,
  value: T
) => U;

export type ImageCallback<U = any> = (
  this: Jimp,
  err: Error | null,
  value: Jimp,
  coords: {
    x: number;
    y: number;
  }
) => U;

type BlendMode = {
  mode: string;
  opacitySource: number;
  opacityDest: number;
};

type ChangeName = 'background' | 'scan' | 'crop';

type ListenableName =
  | 'any'
  | 'initialized'
  | 'before-change'
  | 'changed'
  | 'before-clone'
  | 'cloned'
  | ChangeName;

type ListenerData<T extends ListenableName> = T extends 'any'
  ? any
  : T extends ChangeName
    ? {
      eventName: 'before-change' | 'changed';
      methodName: T;
      [key: string]: any;
    }
    : {
      eventName: T;
      methodName: T extends 'initialized'
        ? 'constructor'
        : T extends 'before-change' | 'changed'
          ? ChangeName
          : T extends 'before-clone' | 'cloned' ? 'clone' : any;
    };

type URLOptions = {
  url: string;
  compression?: boolean;
  headers: {
    [key: string]: any;
  };
};

export interface Bitmap {
  data: Buffer;
  width: number;
  height: number;
}

export interface RGB {
  r: number;
  g: number;
  b: number;
}

export interface RGBA {
  r: number;
  g: number;
  b: number;
  a: number;
}

export default Jimp;
