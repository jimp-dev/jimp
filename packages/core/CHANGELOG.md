# v0.16.3 (Sat Feb 04 2023)

#### üêõ Bug Fix

- Fix edgeHandling types [#1080](https://github.com/jimp-dev/jimp/pull/1080) ([@domdomegg](https://github.com/domdomegg))
- upgrade file-type [#1108](https://github.com/jimp-dev/jimp/pull/1108) ([@krudos](https://github.com/krudos))

#### ‚ö†Ô∏è Pushed to `master`

- try this ([@hipstersmoothie](https://github.com/hipstersmoothie))
- update linting ([@hipstersmoothie](https://github.com/hipstersmoothie))
- upgrade prettier ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 3

- [@krudos](https://github.com/krudos)
- Adam Jones ([@domdomegg](https://github.com/domdomegg))
- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v0.14.0 (Mon Jun 29 2020)

#### üöÄ Enhancement

- include Addition (Add) blending mode + Officially drop support for Node 8 [#904](https://github.com/oliver-moran/jimp/pull/904) ([@GlitchyPSIX](https://github.com/GlitchyPSIX))

#### Authors: 1

- GlitchyPSI ([@GlitchyPSIX](https://github.com/GlitchyPSIX))

---

# v0.11.0 (Fri May 15 2020)

#### üöÄ Enhancement

- Removed Core-JS as a dependency. [#882](https://github.com/oliver-moran/jimp/pull/882) ([@EricRabil](https://github.com/EricRabil))

#### üêõ Bug Fix

- Make callback optional for Jimp.rgbaToInt [#889](https://github.com/oliver-moran/jimp/pull/889) ([@HanKruiger](https://github.com/HanKruiger))

#### Authors: 2

- Eric Rabil ([@EricRabil](https://github.com/EricRabil))
- Han Kruiger ([@HanKruiger](https://github.com/HanKruiger))

---

# v0.10.2 (Tue Apr 14 2020)

#### üêõ Bug Fix

- Rewrite handling EXIF orientation ‚Äî add tests, make it plugin-independent [#875](https://github.com/oliver-moran/jimp/pull/875) ([@skalee](https://github.com/skalee))

#### Authors: 1

- Sebastian Ska≈Çacki ([@skalee](https://github.com/skalee))

---

# v0.10.0 (Mon Mar 30 2020)

#### üöÄ Enhancement

- Properly split constructor and instance types [#867](https://github.com/oliver-moran/jimp/pull/867) ([@forivall](https://github.com/forivall))

#### Authors: 1

- Emily Marigold Klassen ([@forivall](https://github.com/forivall))

---

# v0.9.6 (Wed Mar 18 2020)

#### üêõ Bug Fix

- Relax mkdirp dependency to allow newer minimist [#857](https://github.com/oliver-moran/jimp/pull/857) ([@Den-dp](https://github.com/Den-dp))

#### üè† Internal

- Fix TypeScript error on 'next' [#858](https://github.com/oliver-moran/jimp/pull/858) ([@crutchcorn](https://github.com/crutchcorn))

#### Authors: 2

- Corbin Crutchley ([@crutchcorn](https://github.com/crutchcorn))
- Denis Bendrikov ([@Den-dp](https://github.com/Den-dp))

---

# v0.9.3 (Tue Nov 26 2019)

#### üêõ Bug Fix

- `@jimp/cli`, `@jimp/core`, `@jimp/custom`, `jimp`, `@jimp/plugin-blit`, `@jimp/plugin-blur`, `@jimp/plugin-circle`, `@jimp/plugin-color`, `@jimp/plugin-contain`, `@jimp/plugin-cover`, `@jimp/plugin-crop`, `@jimp/plugin-displace`, `@jimp/plugin-dither`, `@jimp/plugin-fisheye`, `@jimp/plugin-flip`, `@jimp/plugin-gaussian`, `@jimp/plugin-invert`, `@jimp/plugin-mask`, `@jimp/plugin-normalize`, `@jimp/plugin-print`, `@jimp/plugin-resize`, `@jimp/plugin-rotate`, `@jimp/plugin-scale`, `@jimp/plugin-shadow`, `@jimp/plugin-threshold`, `@jimp/plugins`, `@jimp/test-utils`, `@jimp/bmp`, `@jimp/gif`, `@jimp/jpeg`, `@jimp/png`, `@jimp/tiff`, `@jimp/types`, `@jimp/utils`
  - Fix regeneratorRuntime errors [#815](https://github.com/oliver-moran/jimp/pull/815) ([@crutchcorn](https://github.com/crutchcorn) [@hipstersmoothie](https://github.com/hipstersmoothie))
- `@jimp/core`
  - Follow redirects [#789](https://github.com/oliver-moran/jimp/pull/789) ([@SaWey](https://github.com/SaWey) sander@solora.be)

#### Authors: 2

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Corbin Crutchley ([@crutchcorn](https://github.com/crutchcorn))
