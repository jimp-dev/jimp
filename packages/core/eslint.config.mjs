import shared from "@jimp/config-eslint/base.js";
export default [
  ...shared,
  {
    rules: {
      "@typescript-eslint/no-explicit-any": "off",
    },
  },
];
