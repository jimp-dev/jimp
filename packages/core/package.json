{
  "name": "@jimp/core",
  "version": "0.3.9",
  "description": "Jimp core",
  "main": "dist/index.js",
  "module": "es/index.js",
  "files": [
    "dist",
    "es",
    "jimp.d.ts",
    "fonts"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/oliver-moran/jimp.git"
  },
  "bugs": {
    "url": "https://github.com/oliver-moran/jimp/issues"
  },
  "scripts": {
    "test": "cross-env BABEL_ENV=test mocha --require @babel/register test/**/*.js",
    "test:watch": "npm run test -- --reporter min --watch",
    "test:coverage": "nyc npm run test",
    "build": "npm run build:node:production && npm run build:module",
    "build:watch": "npm run build:node:debug -- -- --watch --verbose",
    "build:debug": "npm run build:node:debug",
    "build:module": "cross-env BABEL_ENV=module babel src -d es --source-maps --config-file ../../babel.config.js",
    "build:node": "babel src -d dist --source-maps --config-file ../../babel.config.js",
    "build:node:debug": "cross-env BABEL_ENV=development npm run build:node",
    "build:node:production": "cross-env BABEL_ENV=production npm run build:node"
  },
  "author": "Oliver Moran <oliver.moran@gmail.com>",
  "license": "MIT",
  "dependencies": {
    "@jimp/utils": "^0.3.9",
    "any-base": "^1.1.0",
    "buffer": "^5.2.0",
    "core-js": "^2.5.7",
    "exif-parser": "^0.1.12",
    "file-type": "^9.0.0",
    "load-bmfont": "^1.3.1",
    "mkdirp": "0.5.1",
    "phin": "^2.9.1",
    "pixelmatch": "^4.0.2",
    "tinycolor2": "^1.4.1"
  },
  "devDependencies": {
    "should": "^13.2.3"
  },
  "xo": false,
  "publishConfig": {
    "access": "public"
  }
}
