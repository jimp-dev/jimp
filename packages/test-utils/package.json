{
  "name": "@jimp/test-utils",
  "version": "0.22.11",
  "description": "Test utils for jimp extensions.",
  "main": "dist/index.js",
  "repository": "jimp-dev/jimp",
  "module": "es/index.js",
  "scripts": {
    "test": "cross-env BABEL_ENV=test mocha --require @babel/register --recursive test  --extension js",
    "test:watch": "npm run test -- --reporter min --watch",
    "test:coverage": "nyc npm run test",
    "build": "npm run build:node:production && npm run build:module",
    "build:watch": "npm run build:node:debug -- -- --watch --verbose",
    "build:debug": "npm run build:node:debug",
    "build:module": "cross-env BABEL_ENV=module babel src -d es --source-maps --config-file ../../babel.config.js",
    "build:node": "babel src -d dist --source-maps --config-file ../../babel.config.js",
    "build:node:debug": "cross-env BABEL_ENV=development npm run build:node",
    "build:node:production": "cross-env BABEL_ENV=production npm run build:node"
  },
  "author": "",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "pngjs": "^6.0.0"
  },
  "devDependencies": {
    "@jimp/custom": "link:../custom"
  }
}
