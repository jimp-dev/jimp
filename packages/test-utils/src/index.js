import should from 'should'; // Ensure should to load in browser through browserify.

export const Jimp = require('./jgd-wrapper');

// eslint-disable-next-line no-use-extend-native/no-use-extend-native
const shouldAssertion = {}.should.be.constructor.prototype;

export function hasOwnProp(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}

export function hashForEach(hash, func) {
  for (const key in hash) if (hasOwnProp(hash, key)) func(key, hash[key]);
}

export function getTestDir(dir) {
  const testRE = /\/[^/]+\.test\.js($|\?.*)/;
  if (
    typeof document !== 'undefined' &&
    document &&
    document.getElementsByTagName
  ) {
    const scripts = document.getElementsByTagName('script');

    const testScript = [...scripts].find(
      script => script.src.match(testRE) && script.src.includes(dir)
    );

    if (testScript) {
      return testScript.src.replace(testRE, '');
    }

    throw new TypeError('Cant discover the web test directory');
  } else {
    if (typeof dir === 'undefined') {
      throw new TypeError('Cant discover the env test directory');
    }

    return dir;
  }
}

export function isWeb(warn) {
  if (typeof window !== 'undefined' && window.document) {
    console.warn(warn);
    return true;
  }

  return false;
}

const sup = '⁰¹²³⁴⁵⁶⁷⁸⁹ᵃᵇᶜᵈᵉᶠ';

export function jgdReadableMatrix(img) {
  const rMatrix = [];
  let line = [];
  const len = img.data.length;

  for (let i = 0; i < len; i++) {
    let pix = img.data[i].toString(16).toUpperCase();

    while (pix.length < 8) pix = '0' + pix;

    line.push(
      pix.replace(/(..)(..)(..)(.)(.)/, (sel, r, g, b, a1, a2) => {
        const a = sup[parseInt(a1, 16)] + sup[parseInt(a2, 16)];
        return r + '-' + g + '-' + b + a;
      })
    );
    if (i > 0 && (i + 1) % img.width === 0) {
      rMatrix.push(line.join(' '));
      line = [];
    }
  }

  return rMatrix.join('\n');
}

shouldAssertion.sameJGD = function(targetJGD, message) {
  message = message ? ' ' + message : '';
  const testJGD = this.obj;
  should.exist(testJGD.width, 'Width was not defined.' + message);
  should.exist(testJGD.height, 'Height was not defined.' + message);
  testJGD.width.should.be.equal(
    targetJGD.width,
    'Width is not the expected.' + message
  );
  testJGD.height.should.be.equal(
    targetJGD.height,
    'Height is not the expected.' + message
  );
  const matrixMsg = message || 'The pixel matrix is not the expected.';
  jgdReadableMatrix(testJGD).should.be.equal(
    jgdReadableMatrix(targetJGD),
    matrixMsg
  );
};

export function donutJGD(_, i, X) {
  /* eslint comma-spacing: off */
  return {
    width: 10,
    height: 10,
    data: [
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      i,
      X,
      X,
      i,
      _,
      _,
      _,
      _,
      _,
      X,
      X,
      X,
      X,
      X,
      X,
      _,
      _,
      _,
      i,
      X,
      X,
      i,
      i,
      X,
      X,
      i,
      _,
      _,
      X,
      X,
      i,
      _,
      _,
      i,
      X,
      X,
      _,
      _,
      X,
      X,
      i,
      _,
      _,
      i,
      X,
      X,
      _,
      _,
      i,
      X,
      X,
      i,
      i,
      X,
      X,
      i,
      _,
      _,
      _,
      X,
      X,
      X,
      X,
      X,
      X,
      _,
      _,
      _,
      _,
      _,
      i,
      X,
      X,
      i,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _,
      _
    ]
  };
}

const colors = {
  '▴': 0xff0000ff, // Red
  '▵': 0xff00007f, // Red half-alpha
  '▸': 0x00ff00ff, // Green
  '▹': 0x00ff007f, // Green half-alpha
  '▾': 0x0000ffff, // Blue
  '▿': 0x0000ff7f, // Blue half-alpha
  '◆': 0xffff00ff, // Yellow
  '◇': 0xffff007f, // Yellow half-alpha
  '▪': 0x00ffffff, // Cyan
  '▫': 0x00ffff7f, // Cyan half-alpha
  '▰': 0xff00ffff, // Magenta
  '▱': 0xff00ff7f, // Magenta half-alpha
  ' ': 0x00000000, // Transparent black
  '■': 0x000000ff, // Black
  '0': 0x000000ff, // Black
  '1': 0x111111ff,
  '2': 0x222222ff,
  '3': 0x333333ff,
  '▩': 0x404040ff, // Dark gray (1/4 white)
  '4': 0x444444ff,
  '5': 0x555555ff,
  '6': 0x666666ff,
  '7': 0x777777ff,
  '8': 0x888888ff,
  '▦': 0x808080ff, // Half gray (1/2 white)
  '9': 0x999999ff,
  A: 0xaaaaaaff,
  B: 0xbbbbbbff,
  '▥': 0xbfbfbfff, // Light gray (3/4 white)
  C: 0xccccccff,
  D: 0xddddddff,
  E: 0xeeeeeeff,
  F: 0xffffffff, // White
  '□': 0xffffffff // White
};

function throwUndefinedChar(char) {
  const cList = [];
  hashForEach(colors, (k, c) => {
    c = c.toString(16);
    while (c.length < 8) c = '0' + c;
    cList.push(k + '=' + c);
  });
  throw new Error(
    'The char "' +
      char +
      '" do not defines a color. ' +
      'This are the valid chars: ' +
      cList.join(' ')
  );
}

/* Build a JGD object from a list of strings */
export function mkJGD(...args) {
  const jgd = { width: 0, height: args.length, data: [] };
  for (let y = 0; y < jgd.height; y++) {
    const line = args[y];
    jgd.width = line.length;
    const w = jgd.width;

    for (let x = 0; x < w; x++) {
      if (typeof colors[line[x]] === 'undefined') {
        throwUndefinedChar(line[x]);
      } else {
        jgd.data[y * w + x] = colors[line[x]];
      }
    }
  }
  return jgd;
}

// Helps to debug image data
export function jgdToStr(jgd) {
  const colors2 = {};
  hashForEach(colors, (k, c) => {
    colors2[c] = k;
  });
  const lines = [];
  const w = jgd.width;
  for (let y = 0; y < jgd.height; y++) {
    lines[y] = '';
    for (let x = 0; x < w; x++) {
      const k = colors2[jgd.data[y * w + x]] || '?';
      lines[y] += k;
    }
  }
  return lines.map(l => "'" + l + "'").join('\n');
}
