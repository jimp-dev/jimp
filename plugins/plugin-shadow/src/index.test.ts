import { expect, test, describe } from "vitest";

import { makeTestImage } from "@jimp/test-utils";

import { methods } from "./index.js";
import { createJimp } from "@jimp/core";

const jimp = createJimp({ plugins: [methods] });

describe("Shadow", () => {
  test("creates a shadow", async () => {
    const testImage = jimp.fromBitmap(
      makeTestImage(
        "          ",
        "    ◆◆    ",
        "   ◆▦▦◆   ",
        "  ◆▦▦▦▦◆  ",
        "   ◆▦▦◆   ",
        "    ◆◆    ",
        "          ",
      ),
    );

    expect(testImage.shadow({ x: -1, y: 1, blur: 1 })).toMatchSnapshot();
  });
});
