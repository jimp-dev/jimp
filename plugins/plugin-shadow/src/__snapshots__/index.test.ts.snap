// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shadow > creates a shadow 1`] = `
Visualization:

          
 0▱▥1▪◇▫  
▪ED▥93F5▫ 
▰C7▸◇▥▵□A1
F◇▦▾2A651▫
◆■B■■3F6▰▰
E▱▥!▸0■D6▫

Data:

00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰ 00-00-00⁰⁰
00-00-00⁰⁰ 04-04-00⁰² 0F-0F-00⁰ᵇ 20-20-02¹⁷ E4-E4-00ᶠᶠ E5-E5-00ᶠᶠ 18-18-01¹² 09-09-00⁰⁶ 00-00-00⁰⁰ 00-00-00⁰⁰
03-03-00⁰² 0D-0D-01⁰ᵃ 22-22-05¹ᶜ D8-D8-02ᶠᶠ 73-73-63ᶠᶠ 72-72-64ᶠᶠ DD-DD-01ᶠᶠ 18-18-03¹³ 07-07-00⁰⁵ 00-00-00⁰⁰
0B-0B-00⁰⁸ 20-20-05¹ᵃ D4-D4-03ᶠᶠ 75-75-5Dᶠᶠ 7C-7C-59ᶠᶠ 7A-7A-5Aᶠᶠ 73-73-60ᶠᶠ DB-DB-01ᶠᶠ 16-16-02¹⁰ 06-06-00⁰⁴
13-13-01⁰ᵉ 33-33-08²ᵃ 5D-5D-1A⁵⁴ C0-C0-19ᶠᶠ 89-89-5Dᶠᶠ 87-87-5Cᶠᶠ C2-C2-12ᶠᶠ 4E-4E-11⁴³ 25-25-04¹ᶜ 0B-0B-00⁰⁷
16-16-01¹⁰ 3A-3A-0A²ᶠ 65-65-1D⁵ᶜ 87-87-3A⁸⁷ BE-BE-30ᶠᶠ BE-BE-2Bᶠᶠ 7F-7F-2F⁷ᵃ 56-56-14⁴ᵃ 2A-2A-04²⁰ 0D-0D-00⁰⁹
0F-0F-01⁰ᵇ 2C-2C-07²⁴ 54-54-16⁴ᵃ 78-78-2D⁷³ 8B-8B-3C⁸ᵇ 87-87-38⁸⁶ 6E-6E-24⁶⁶ 45-45-0F³ᵃ 1E-1E-03¹⁷ 09-09-00⁰⁶
`;
